### PR TITLE
(FACT-2746) Added cloud resolver

### PR DESCRIPTION
**Pull request by sebastian-miclea**

### CODE REVIEW

File: lib/facter/resolvers/cloud.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+module Facter
+  module Resolvers
+    class Cloud < BaseResolver
+      # cloud_provider
+
+      @semaphore = Mutex.new
```
Comment (Filipovici-Andrei) [inline lib/facter/resolvers/cloud.rb:8]: We are no longer using the @semaphore

File: spec/facter/resolvers/cloud_spec.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    before do
+      allow(File).to receive(:directory?).with('/var/lib/dhcp').and_return(true)
+      allow(Facter::Util::FileHelper)
+        .to receive(:safe_readlines)
+        .with('/var/lib/dhcp/dhclient.rhel.lease.8')
+        .and_return(content)
+      allow(Dir).to receive(:entries).with('/var/lib/dhcp').and_return(['.', 'dhclient.rhel.lease.8'])
+    end
+
+    it 'returns azure' do
```
Comment (Filipovici-Andrei) [inline spec/facter/resolvers/cloud_spec.rb:58]: it 'returns nil'

File: spec/facter/resolvers/cloud_spec.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      expect(cloud_resolver.resolve(:cloud_provider)).to eq('azure')
+    end
+  end
+
+  context 'when lease file is found without option 245' do
```
Comment (Filipovici-Andrei) [inline spec/facter/resolvers/cloud_spec.rb:46]: Could you add a test when the lease file is found but it can't be read?

**Review by Filipovici-Andrei — CHANGES_REQUESTED** [review]

File: spec/facter/resolvers/cloud_spec.rb (hunk 3 above)
Comment (sebastian-miclea) [reply spec/facter/resolvers/cloud_spec.rb:46]: Yes, updated.

**Review by Filipovici-Andrei — APPROVED** [review]

**Pull request merged**